features/22-email-verification/implementation.md

```diff
@@ -0,0 +1,216 @@
+is returned if the user is not verified.
+
+This preprocessor is to be added to all handlers that may require verification.
+This preprocessor is also inserted to plugin handlers and lambdas that
+supply `verify_required` in registration info.
+
+### Database
+
+The following columns are to be added to `_auth`:
+
+* `verified` (boolean)
+  Indicates whether the user is verified
+
+* `verified_keys` (JSON, string: any)
+  Indicates which record key is marked as verified
+
+New table `_auth_challenge` for storing verification code.
+
+* auth_id (string)
+  The ID of the user who requested the challenge.
+
+* record_key (string)
+  The record key of the information that requires verification.
+
+* record_value (string)
+  The value for the record key that requires verification.
+
+* code (string)
+  The verification challenge.
+
+* consumed (boolean)
+  Whether the challenge is challenge. A consumed challenge cannot be
+  reused.
+
+* created_at (datetime)
+  The date/time when the challenge is created.
+
+* expire_at (datetime, optional)
+  The date/time when the challenge will expire.
+
+### Configuration
+
+* `VERIFY_KEYS` (array, string)
+  Which piece of auth record key is enabled for verification. Can be `email`,
+  `phone` or `email,phone`. The specified auth record key must also exist in the
+  `AUTH_RECORD_KEYS`.
+
+* `VERIFY_REQUIRED` (boolean)
+  Whether verification is required. When this is true, verification email/SMS
+  will be sent upon signup. The user must be verified before they can call
+  other APIs.
+
+* `VERIFY_CRITERIA` (string, optional, default `any`)
+  Either `any` or `all`. If `any`, the user is verified if any auth keys
+  are verified. If `all`, the user is verified if all auth keys are verified.
+
+* `VERIFY_KEYS_EMAIL`, `VERIFY_KEYS_PHONE`
+
+  * `_REQUIRED` (boolean)
+
+  * `_CODE_FORMAT` (string), `numeric` or `complex`
+
+  * `_PROVIDER` (string, the provider of the verification)
+    For email, it can be `smtp`. For phone, it can be SMS gateways.
+
+  * `_PROVIDER_*` provider specific configuration.
+
+## SDK
+
+### AuthContainer
+
+The AuthContainer should add these properties which should be updated
+upon receiving auth response from the server.
+
+* `verified` (boolean) whether the user is verified
+* `verifiedRecordKeys` (object, string: any) which piece of information is verified
+
+### APIs
+
+#### Phone
+
+```
+skygear.auth.signupWithPhone("+85221559299", "passw0rd")
+skygear.auth.sendPhoneVerification('+85221559299')
+skygear.auth.verifyPhone('+85221559299', '123456')
+```
+
+#### Email
+
+```
+skygear.auth.signupWithEmail("johndoe@example.com", "passw0rd")
+skygear.auth.sendEmailVerification('johndoe@example.com')
+skygear.auth.verifyEmail('johndoe@example.com', '123456')
+```
+
+
+## Plugin
+
+Handler and Lambda will be allowed to register a handler or a lambda
+that requires user to be verified.
+
+```python
+@skygear.handler(name, verify_required=True)
```

Comment:
For the current `user_reqired`, I think it should also be rejected, if `VERIFY_REQUIRED` and the user is not verified?